### ISSUE TITLE
Scrapy vs Django + Django Q + Gunicorn

### ISSUE THREAD

**Issue by iagent** [scrapy/scrapy#2936]
I am currently using scrapy for an application involving **broad crawling.** 

One of my teammates contested to use a combination of django + gunicorn in place of Scrapy as most of the team is conversant with Django and its ORM. The number of websites involved are to the tune of 100K. He suggests to use Gunicorn if we want concurrency and Django Q for queueing requests.

I am not comfortable with the idea as Django and Scrapy are built for completely different problems but not sure how can I make a strong argument.

Could you help me out here? My understanding is that queueing at each stage of crawling and scraping sort of helps in performance. As Scrapy takes care of each stage, even if we temporarily drop performance concern, wouldn't it involve more code writing?

**Comment by ajpen** [comment]
From what I understand, Scrapyd, might more suit to what you're trying to do. http://scrapyd.readthedocs.io/en/stable/
You can use your Django application to schedule crawls to scrapyd. 

There's also the cloud solution available https://scrapinghub.com/scrapy-cloud

**Comment by IAlwaysBeCoding** [comment]
@iagent If you want to see how much work you will be doing, then I suggest you look at the Download Middlewares and the core folder. Everything there will be what you will be needing to core from scratch. From the outlook it looks like Scrapy is just a queue that dispatches Requests and does parsing, but it actually does a few other things.

Your Django and gunicorn setup will cover around 80% of what Scrapy does, but the last 20% will be very important which you will need to implement yourself. 

I've personally thought it over about implementing a new scraping framework using python 3.5+ async or just extend Scrapy via a custom scheduler, middlewares, extensions. Still undecided, even though I have personally created more than 80 Scrapy spiders. 

To be honest, I might contribute more to Scrapy via other tools/utilities that make developing faster. Developing in Scrapy unfortunately is slow as hell once you get into full featured scrapers and beyond simple scrapers that you find on tutorials. The main reason for that is that Scrapy doesn't have a huge ecosystem of sharing like web frameworks such as Django(python), Ruby on Rails(ruby), Laravel(php), Express(nodejs). 

You end up with a lot of people building their own Spiders instead of sharing Item Pipelines, Spider Middlewares, Download Middlewars, Extensions etc. Those are what make the developing of Scrapy much faster. 

In short, to develop faster you need the modules/packages/libraries aka Item Pipelines, Spider Middlewares, Download Middlewars, Extensions to build your scrapers faster.

**Comment by kmike** [comment]
@iagent one can build a successful system with both approaches. I personally would suggest Scrapy, as it handles a lot of edge cases with web crawling; likely it would be faster, more correct and more reliable than a cusom-built solution. Example of issues you'll face if you go with gunicorn + requests:

* requests.get(url).text is not using `<meta>` tags in HTML source code, so encoding would be often incorrect if you're crawling international websites (e.g. on ~10% of Russian websites according to my tests).
* most url-processing libraries are not handling non-ascii URLs properly, i.e. like a browser.
* you will need to be careful when building absolute URLs from extracted relative URLs - e.g. take `<base>` HTML tag in account.

You will also need to handle retries, concurrency limits, etc. yourselves.

Anyways, I think a better place to discuss it is https://www.reddit.com/r/scrapy/, so I'm closing this ticket.
